README.md
# jmbg-validator
(README in English available below)

Program validira i ispisuje osnovne podatke o [JMBGu](https://sh.wikipedia.org/wiki/Jedinstveni_mati%C4%8Dni_broj_gra%C4%91ana).

## Privatnost
Program **ne čuva** JMBG ni u kakvoj bazi niti ga šalje trećim stranama.

## Instalacija
```
pip install jmbg-validator
```

Dolazi sa alatom za komandnu liniju (CLI) po imenu `jmbg-validator` i Python modulom `jmbg_validator`

## Upotreba
Matični broj `1234567890123` nije validan i koristi se samo kao primer.

### U komandnoj liniji (CLI)
Alat za komandnu liniju se može koristiti na sledeći način:
```
$ jmbg-validator 1234567890123
Pol                     Validan         Datum rođenja           Region
Muški                   Da              11.12.1989.             Vojvodina, Novi Sad

$ jmbg-validator 1234567890123 -o json
{"pol": "ženski", "datum_rodjenja": "06.09.1992.", "validan": true, "region": "Bosna i Hercegovina, Tuzla"}
```

Pregled svih dostupnih opcija:
```
usage: jmbg-validator [-h] [--output {table,text,json}] jmbg

Validira i ispisuje osnovne podatke o JMBGu

positional arguments:
  jmbg                  13 karaktera JMBGa

optional arguments:
  -h, --help            show this help message and exit
  --output {table,text,json}, -o {table,text,json}
```

### Kao Python modul

```
>>> from jmbg_validator import validate
>>> validate("1234567890123")
JMBG(pol='muški', datum_rodjenja='11.12.1989.', validan=True, region='Vojvodina, Novi Sad')
>>> print(validate("1234567890123"))
Pol: Muški, Validan: Da, Datum rođenja: 11.12.1989., Region: Vojvodina, Novi Sad
>>> validate("1234567890123")
>>> rezultat = JMBG(pol='muški', datum_rodjenja='11.12.1989.', validan=True, region='Vojvodina, Novi Sad')
>>> rezultat.validan
True
```

## Za developere
```
$ python -m venv venv
$ source venv/bin/activate
$ pip install -e .
$ pip install -r dev_requirements.txt
```

### Testiranje
```
$ python -m unittest discover
....................................................................
----------------------------------------------------------------------
Ran 68 tests in 0.001s

OK
```

## Licenca

[Mozilla Public License 2.0](https://www.mozilla.org/en-US/MPL/2.0/)
ili
[ukratko](https://www.tldrlegal.com/l/mpl-2.0)

---

# jmbg-validator

This program validates and displays basic [UMCN](https://en.wikipedia.org/wiki/Unique_Master_Citizen_Number) data. UMCN
was/still is in use in former Yugoslav republics as a way to uniquely identify citizens.

## Privacy
This program **does not** store the UMCN and it **does not** share it with 3rd parties.

## Installation
```
pip install jmbg-validator
```

Comes with a CLI tool `jmbg-validator` and a Python module named `jmbg_validator`

## Usage
UMCN `1234567890123` used below is not valid and is used for demonstration purposes only.

### As a CLI tool
CLI tool can be used in this way:
```
$ jmbg-validator 1234567890123
Pol                     Validan         Datum rođenja           Region
Muški                   Da              11.12.1989.             Vojvodina, Novi Sad

$ jmbg-validator 1234567890123 -o json
{"pol": "ženski", "datum_rodjenja": "06.09.1992.", "validan": true, "region": "Bosna i Hercegovina, Tuzla"}
```

Overview of available arguments:
```
usage: jmbg-validator [-h] [--output {table,text,json}] jmbg

Validira i ispisuje osnovne podatke o JMBGu

positional arguments:
  jmbg                  13 karaktera JMBGa

optional arguments:
  -h, --help            show this help message and exit
  --output {table,text,json}, -o {table,text,json}
```

### As a Python module

```
>>> from jmbg_validator import validate
>>> validate("1234567890123")
JMBG(pol='muški', datum_rodjenja='11.12.1989.', validan=True, region='Vojvodina, Novi Sad')
>>> print(validate("1234567890123"))
Pol: Muški, Validan: Da, Datum rođenja: 11.12.1989., Region: Vojvodina, Novi Sad
>>> validate("1234567890123")
>>> rezultat = JMBG(pol='muški', datum_rodjenja='11.12.1989.', validan=True, region='Vojvodina, Novi Sad')
>>> rezultat.validan
True
```

## For developers
```
$ python -m venv venv
$ source venv/bin/activate
$ pip install -e .
$ pip install -r dev_requirements.txt
```

### Running tests
```
$ python -m unittest discover
....................................................................
----------------------------------------------------------------------
Ran 68 tests in 0.001s

OK
```

## License

[Mozilla Public License 2.0](https://www.mozilla.org/en-US/MPL/2.0/)
or
[in short](https://www.tldrlegal.com/l/mpl-2.0)

CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.0.2 (2021-10-30)
### Fix
* Includes long and short descriptions for PyPI ([`d194462`](https://github.com/ivica-k/jmbg-validator/commit/d194462ff4bc69510c7c8d924208ac34ddb7ee16))

## v1.0.1 (2021-10-30)
### Fix
* Configures GH action to release to PyPI ([`1f872e9`](https://github.com/ivica-k/jmbg-validator/commit/1f872e99294faba898bbe8c36208d885f8dddb8f))

## v1.0.0 (2021-10-30)
### Documentation
* Improves the README file; adds English version ([`5c54d1e`](https://github.com/ivica-k/jmbg-validator/commit/5c54d1eae68a6b34091db846adc3f3ec4367aa9b))


tests/test_jmbg_validator.py
import unittest

from parameterized import parameterized

from jmbg_validator.exceptions import InvalidJMBG, InvalidJMBGLength, InvalidDateException, NotANumber, InvalidRegion
from jmbg_validator.main import (
    verify_length,
    verify_is_number,
    verify_date,
    split_fields,
    get_region,
    get_sex,
    get_dob,
    get_sum,
    verify_control,
)


class TestJMBGValidator(unittest.TestCase):
    def setUp(self):
        pass

    @parameterized.expand(
        [
            ("1234567890123",),
            ("3210987654321",),
        ]
    )
    def test_verify_length_pass(self, input_string):
        actual_result = verify_length(input_string)

        self.assertTrue(actual_result)

    @parameterized.expand(
        [
            ("123456",),
            ("32109876543212",),
            (32109876543212,),
        ]
    )
    def test_verify_length_fail(self, input_string):
        self.assertRaises(InvalidJMBGLength, verify_length, input_string)

    @parameterized.expand(
        [
            ("123456",),
            ("32109876543212",),
            ("400",),
        ]
    )
    def test_verify_is_number_pass(self, input_string):
        actual_result = verify_is_number(input_string)

        self.assertTrue(actual_result)

    @parameterized.expand(
        [
            ("123456a",),
            ("32109876543212k",),
            ("jmbg",),
            ("Ӝ",),
        ]
    )
    def test_verify_is_number_fail(self, input_string):
        self.assertRaises(NotANumber, verify_is_number, input_string)

    @parameterized.expand(
        [
            (28, 10, 1986),
            (4, 4, 2018),
            (12, 12, 1974),
            (31, 12, 1822),
            (29, 12, 1222),
            (28, 2, 1679),
        ]
    )
    def test_verify_date_pass(self, day, month, year):
        actual_result = verify_date(day, month, year)

        self.assertTrue(actual_result)

    @parameterized.expand(
        [
            (30, 102, 1986),
            (4, 13, 2018),
            (32, 12, 1974),
            (31, 12, 12345),
            (29, 12, "aaaa"),
            (28, "bbbb", 1679),
        ]
    )
    def test_verify_date_fail(self, day, month, year):
        self.assertRaises(InvalidDateException, verify_date, day, month, year)

    @parameterized.expand(
        [
            ("1234567890123", ("12", "34", "567", "89", "012", "3")),
            ("3210987654321", ("32", "10", "987", "65", "432", "1")),
        ]
    )
    def test_split_fields_pass(self, input_string, expected_result):
        actual_result = split_fields(input_string)

        self.assertEqual(actual_result, expected_result)

    @parameterized.expand(
        [
            ("abcdefghjklmn",),
            ("321098765432",),
            ("12345",),
        ]
    )
    def test_split_fields_fail(self, input_string):
        self.assertRaises(InvalidJMBG, split_fields, input_string)

    @parameterized.expand(
        [
            ("89", "Vojvodina, Sremska Mitrovica"),
            ("47", "Severna Makedonija, Tetovo"),
            ("27", "Crna Gora, Berane, Rožaje, Plav, Andrijevica"),
            ("08", "stranci u Vojvodini"),
        ]
    )
    def test_get_region_pass(self, input_string, expected_result):
        actual_result = get_region(input_string)

        self.assertEqual(actual_result, expected_result)

    @parameterized.expand(
        [
            ("97",),
            ("40",),
            ("97",),
            ("54",),
            ("20",),
            ("90",),
        ]
    )
    def test_get_region_fail(self, input_string):
        self.assertRaises(InvalidRegion, get_region, input_string)

    @parameterized.expand(
        [
            ("212", "muški"),
            ("499", "muški"),
            ("001", "muški"),
            ("630", "ženski"),
            ("999", "ženski"),
            ("500", "ženski"),
        ]
    )
    def test_get_sex_pass(self, input_string, expected_result):
        actual_result = get_sex(input_string)

        self.assertEqual(actual_result, expected_result)

    @parameterized.expand(
        [
            ("-970",),
            ("1040",),
            ("2222",),
            ("98123",),
            ("a",),
        ]
    )
    def test_get_sex_fail(self, input_string):
        self.assertRaises(Exception, get_sex, input_string)

    @parameterized.expand(
        [
            (24, 10, 1973, "24.10.1973."),
            (5, 12, 2012, "05.12.2012."),
            (30, 3, 1990, "30.03.1990."),
            ("24", "10", "1973", "24.10.1973."),
            ("5", 12, 2012, "05.12.2012."),
            (30, "03", 1990, "30.03.1990."),
        ]
    )
    def test_get_dob_pass(self, day, month, year, expected_result):
        actual_result = get_dob(day, month, year)

        self.assertEqual(actual_result, expected_result)

    @parameterized.expand(
        [
            (
                24,
                10,
                19730,
            ),
            (
                24,
                "a",
                999,
            ),
            (
                5,
                132.4,
                "2012",
            ),
            (
                300,
                3.5,
                1990,
            ),
        ]
    )
    def test_get_dob_fail(self, day, month, year):
        self.assertRaises(Exception, get_dob, day, month, year)

    @parameterized.expand(
        [
            ("1234567890123", 226),
            ("2345678921234", 240),
            ("5298529051067", 247),
            ("6339638162178", 252),
        ]
    )
    def test_get_sum_pass(self, input_string, expected_result):
        actual_result = get_sum(input_string)

        self.assertEqual(actual_result, expected_result)

    @parameterized.expand(
        [
            (229, 2),
            (233, 9),
            (213, 7),
            (234, 8),
        ]
    )
    def test_verify_control_pass(self, sum, control):
        self.assertTrue(verify_control(sum, control))


if __name__ == "__main__":
    unittest.main()


jmbg_validator/main.py
#!/usr/bin/env python3

import re
from datetime import datetime
import argparse
from sys import exit

from jmbg_validator.exceptions import (
    InvalidJMBG,
    InvalidJMBGLength,
    InvalidDateException,
    NotANumber,
    InvalidRegion,
)
from jmbg_validator.matrices import REGION_MATRIX
from jmbg_validator.output import JMBG

REGEX = r"(\d{2})(\d{2})(\d{3})(\d{2})(\d{3})(\d{1})"
DATETIME_FORMAT = "%d.%m.%Y."
OUTPUT_CHOICES = ["table", "text", "json"]


def verify_date(day, month, year):
    try:
        if isinstance(day, str):
            day = int(day)

        if isinstance(month, str):
            month = int(month)

        datetime(year, month, day)

        return True

    except (ValueError, TypeError):
        raise InvalidDateException(f"Datum {day}.{month}.{year} nije validan.")


def verify_length(jmbg):
    try:
        if len(jmbg) != 13:
            raise InvalidJMBGLength(f"JMBG mora da ima 13 znakova, ne {len(jmbg)}.")
        else:
            return True
    except TypeError:
        raise InvalidJMBGLength(f"JMBG mora biti string, ne {type(jmbg)}.")


def verify_is_number(jmbg):
    try:
        if not jmbg.isdigit():
            raise NotANumber(f"JMBG ne može da sadrži slova.")

    except AttributeError:
        raise NotANumber(f"JMBG mora biti string, ne {type(jmbg)}.")

    return True


def get_sum(jmbg):
    first = jmbg[0:6]
    second = jmbg[6:-1]
    multiplier = [7, 6, 5, 4, 3, 2]

    first_sum = sum([int(a) * b for a, b in zip(first, multiplier)])
    second_sum = sum([int(a) * b for a, b in zip(second, multiplier)])

    return first_sum + second_sum


def verify_control(jmbg_sum, control):
    control = int(control)
    remainder = jmbg_sum % 11

    if remainder > 1:
        return control == 11 - remainder

    elif remainder == 0 and remainder == control:
        return True

    return False


def split_fields(jmbg):
    match = re.match(REGEX, jmbg)

    if not match:
        raise InvalidJMBG(f"JMBG '{jmbg}' nije ispravan")

    return match.groups()


def get_sex(sex):
    sex = int(sex)

    if 0 <= sex <= 499:
        return "muški"

    elif 500 <= sex <= 999:
        return "ženski"

    else:
        raise Exception(f"Pol mora biti u rasponu od 0-499 i 500-999, ne '{sex}'")


def get_dob(day, month, year):
    return datetime(int(year), int(month), int(day)).strftime(DATETIME_FORMAT)


def get_region(region):
    region = str(region)
    try:
        return REGION_MATRIX[region]

    except KeyError:
        raise InvalidRegion(f"Region '{region}' ne postoji")


def _parse_args():
    parser = argparse.ArgumentParser(description="Validira i ispisuje osnovne podatke o JMBGu")
    parser.add_argument("jmbg", help="13 karaktera JMBGa")
    parser.add_argument(
        "--output",
        "-o",
        default="table",
        required=False,
        choices=OUTPUT_CHOICES,
    )

    return parser.parse_args()


def validate(jmbg):
    if verify_length(jmbg) and verify_is_number(jmbg):
        day, month, year, region, sex, control = split_fields(jmbg)
        human_year = int(f"1{year}")

        if verify_date(day, month, human_year):
            sex = get_sex(sex)
            dob = get_dob(day, month, human_year)
            region = get_region(region)

            jmbg_sum = get_sum(jmbg)
            valid = verify_control(jmbg_sum, control)

            return JMBG(sex, dob, valid, region)


def cli():
    args = _parse_args()

    try:
        if args.output.lower() == "table":
            print(validate(args.jmbg).to_table())

        elif args.output.lower() == "json":
            print(validate(args.jmbg).to_json().decode())

        else:
            print(validate(args.jmbg).to_text())

    except Exception as ex:
        exit(ex)


if __name__ == "__main__":
    cli()


jmbg_validator/matrices.py
REPUBLIC_MATRIX = {
    0: "stranac",
    1: "Bosna i Hercegovina",
    2: "Crna Gora",
    3: "Hrvatska",
    4: "Severna Makedonija",
    5: "Slovenija",
    6: "privremeni boravak",
    7: "uža Srbija",
    8: "Vojvodina",
    9: "Kosovo i Metohija",
}

REGION_MATRIX = {
    "00": "mešana upotreba",
    "01": "stranci u Bosni i Hercegovini",
    "02": "stranci u Crnoj Gori",
    "03": "stranci u Hrvatskoj",
    "04": "stranci u Severnoj Makedoniji",
    "05": "stranci u Sloveniji",
    "07": "stranci u užoj Srbiji",
    "08": "stranci u Vojvodini",
    "09": "stranci na Kosovu i Metohiji",
    "10": f"{REPUBLIC_MATRIX.get(1)}, Banja Luka",
    "11": f"{REPUBLIC_MATRIX.get(1)}, Bihać",
    "12": f"{REPUBLIC_MATRIX.get(1)}, Doboj",
    "13": f"{REPUBLIC_MATRIX.get(1)}, Goražde",
    "14": f"{REPUBLIC_MATRIX.get(1)}, Livno",
    "15": f"{REPUBLIC_MATRIX.get(1)}, Mostar",
    "16": f"{REPUBLIC_MATRIX.get(1)}, Prijedor",
    "17": f"{REPUBLIC_MATRIX.get(1)}, Sarajevo",
    "18": f"{REPUBLIC_MATRIX.get(1)}, Tuzla",
    "19": f"{REPUBLIC_MATRIX.get(1)}, Zenica",
    "21": f"{REPUBLIC_MATRIX.get(2)}, Podgorica",
    "22": f"{REPUBLIC_MATRIX.get(2)}, Bar, Ulcinj",
    "23": f"{REPUBLIC_MATRIX.get(2)}, Budva, Kotor, Tivat",
    "24": f"{REPUBLIC_MATRIX.get(2)}, Herceg Novi",
    "25": f"{REPUBLIC_MATRIX.get(2)}, Cetinje",
    "26": f"{REPUBLIC_MATRIX.get(2)}, Nikšić",
    "27": f"{REPUBLIC_MATRIX.get(2)}, Berane, Rožaje, Plav, Andrijevica",
    "28": f"{REPUBLIC_MATRIX.get(2)}, Bijelo Polje, Mojkovac",
    "29": f"{REPUBLIC_MATRIX.get(2)}, Pljevlja, Žabalj",
    "30": f"{REPUBLIC_MATRIX.get(3)}, Osijek",
    "31": f"{REPUBLIC_MATRIX.get(3)}, Bjelovar, Virovitica, Koprivnica, Pakrac",
    "32": f"{REPUBLIC_MATRIX.get(3)}, Varaždin, Međimurje",
    "33": f"{REPUBLIC_MATRIX.get(3)}, Zagreb",
    "34": f"{REPUBLIC_MATRIX.get(3)}, Karlovac",
    "35": f"{REPUBLIC_MATRIX.get(3)}, Lika",
    "36": f"{REPUBLIC_MATRIX.get(3)}, Istra, Primorje",
    "37": f"{REPUBLIC_MATRIX.get(3)}, Sisak, Banija",
    "38": f"{REPUBLIC_MATRIX.get(3)}, Dalmacija",
    "39": f"{REPUBLIC_MATRIX.get(3)}, mešana upotreba",
    "41": f"{REPUBLIC_MATRIX.get(4)}, Bitolj",
    "42": f"{REPUBLIC_MATRIX.get(4)}, Kumanovo",
    "43": f"{REPUBLIC_MATRIX.get(4)}, Ohrid",
    "44": f"{REPUBLIC_MATRIX.get(4)}, Prilep",
    "45": f"{REPUBLIC_MATRIX.get(4)}, Skoplje",
    "46": f"{REPUBLIC_MATRIX.get(4)}, Strumica",
    "47": f"{REPUBLIC_MATRIX.get(4)}, Tetovo",
    "48": f"{REPUBLIC_MATRIX.get(4)}, Veles",
    "49": f"{REPUBLIC_MATRIX.get(4)}, Štip",
    "50": f"{REPUBLIC_MATRIX.get(5)}",
    "60": f"{REPUBLIC_MATRIX.get(6)}",
    "70": f"{REPUBLIC_MATRIX.get(7)}, Građani upisani u Matičnu knjigu rođenih Diplomatsko Konzularnog Predstavništva Republike Srbije",
    "71": f"{REPUBLIC_MATRIX.get(7)}, Beograd",
    "72": f"{REPUBLIC_MATRIX.get(7)}, Kragujevac, Jagodina",
    "73": f"{REPUBLIC_MATRIX.get(7)}, Niš, Pirot, Toplica",
    "74": f"{REPUBLIC_MATRIX.get(7)}, Leskovac, Vranje",
    "75": f"{REPUBLIC_MATRIX.get(7)}, Zaječar, Bor",
    "76": f"{REPUBLIC_MATRIX.get(7)}, Smederevo, Požarevac",
    "77": f"{REPUBLIC_MATRIX.get(7)}, Mačva, Kolubara",
    "78": f"{REPUBLIC_MATRIX.get(7)}, Čačak, Kraljevo, Kruševac",
    "79": f"{REPUBLIC_MATRIX.get(7)}, Užice",
    "80": f"{REPUBLIC_MATRIX.get(8)}, Novi Sad",
    "81": f"{REPUBLIC_MATRIX.get(8)}, Sombor",
    "82": f"{REPUBLIC_MATRIX.get(8)}, Subotica",
    "83": f"{REPUBLIC_MATRIX.get(8)}, Vrbas",
    "84": f"{REPUBLIC_MATRIX.get(8)}, Kikinda",
    "85": f"{REPUBLIC_MATRIX.get(8)}, Zrenjanin",
    "86": f"{REPUBLIC_MATRIX.get(8)}, Pančevo",
    "87": f"{REPUBLIC_MATRIX.get(8)}, Vršac",
    "88": f"{REPUBLIC_MATRIX.get(8)}, Ruma",
    "89": f"{REPUBLIC_MATRIX.get(8)}, Sremska Mitrovica",
    "91": f"{REPUBLIC_MATRIX.get(9)}, Priština",
    "92": f"{REPUBLIC_MATRIX.get(9)}, Kosovska Mitrovica",
    "93": f"{REPUBLIC_MATRIX.get(9)}, Peć",
    "94": f"{REPUBLIC_MATRIX.get(9)}, Đakovica",
    "95": f"{REPUBLIC_MATRIX.get(9)}, Prizren",
    "96": f"{REPUBLIC_MATRIX.get(9)}, Gnjilane, Kosovska Kamenica, Vitina, Novo Brdo",
}
